scheduler/src/cook/mesos/util.clj

```diff
@@ -552,3 +552,38 @@
   [ch ch-size]
   (->> (repeatedly ch-size #(async/poll! ch))
        (remove nil?)))
+
+(defn reducing-pipe
+  "Reads elements from the `in-chan` channel and supplies elements to the `out-chan` channel.
+   Maintains a state (initialized to `initial-state`) that is updated by applying the reducing function
+   `in-xform` to the current state and the incoming element `(in-xform state element)`.
+   `out-preparer` accepts the current state (i.e. `(out-preparer state)`) and produces `[reset-state out-value]`.
+   When the `out-chan` channel can receive an element, a `out-data-chan` channel is put on the `out-chan`
+   channel. Then the `out-value` is put on this `out-data-chan` and the state is set to the `reset-state`.
+   The optional parameter `on-consumed` can be used to register notifications of when a value has been
+   supplied to the `out-chan` channel.
+
+   Note: This function does not perform error handling, exceptions must be explicitly handled in the
+   provided functions (i.e. in-xform, out-preparer and on-finished). In particular, `out-preparer` must
+   not throw an exception as it will prevent the promise-chan from being fulfilled.
+
+   Note: The roundabout approach of putting a promise-chan on the out-chan before actually placing the
+   value on the promise-chan is necessary as we want to limit calls to `out-preparer` to only when the
+   data is ready to be consumed by `out-chan`."
+  [in-chan in-xform out-chan out-preparer &
```

Comment:
Why don't we call `in-xform` `reducer` and get rid of `out-preparer` entirely?  The `out-preparer` is not core to what is happening here.